src/ripple/app/main/GRPCServer.cpp

```diff
@@ -250,12 +257,44 @@ template <class Request, class Response>
 Role
 GRPCServerImpl::CallData<Request, Response>::getRole()
 {
-    return Role::USER;
+    if (clientIsUnlimited())
+        return Role::IDENTIFIED;
+    else if (wasForwarded())
+        return Role::PROXY;
+    else
+        return Role::USER;
 }
 
 template <class Request, class Response>
-Resource::Consumer
-GRPCServerImpl::CallData<Request, Response>::getUsage()
+bool
+GRPCServerImpl::CallData<Request, Response>::wasForwarded()
```

Comment:
This implementation is different than that of secure_gateway for json-rpc and websocket. Namely, A port becomes UNLIMITED if there is a "User" field in the HTTP header. The presence of an IP address supplied by the client via "Forwarded" or "X-Forwarded-For" headers means that the client of the proxy is charged any fees for usage. This is the role known as "PROXY".

To make grpc consistent, I suggest adding a field in the protocol buffer message, such as "user" when, if set and coming from a secure_gateway host, will make the port UNLIMITED.

Reply:
The way it works right now is there is a field in each protobuf message called `client_ip`. If `client_ip` is set, then the usage fee is charged to the value of `client_ip`. If `client_ip` is not set and the request came from a secure_gateway host, then the client is considered UNLIMITED.
We can add a "User" field to the protobuf message, but I'm not sure how that would actually help or change anything from what we have now. As is, there is already a mechanism for bypassing usage fees via secure_gateway, as well as a mechanism for charging usage fees to the original client ip in the case of proxied RPCs. The role is set for each case. What would adding a "User" field to the message give us that we don't currently have?

Reply:
The User field was intended as a way to track activities in the log file by identified user. As implemented for grpc I left a comment about what happens if the client IP can't be figured out which may lead to a port wrongly being interpreted as from secure_gateway. Other than that, the problem I see is a lack of consistency between grpc & how the other port types behave with secure_gateway. But the intended result for this particular purpose--allowing the reporting server to perform ETL without resource limits--works.

Reply:
What you left in the other comment, about the client IP not being figured out, is a good point actually. Ok, we can add the User field to the gRPC messages then. For when the IP can't be figured out, and for consistency.